### PR TITLE
Install and use shim in the confined snap

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

cc @valentindavid 
cc @xnox

### CODE REVIEW

File: plugins/uefi-capsule/fu-uefi-device.h [hunk 1]
```diff
@@ -66,30 +66,24 @@ typedef enum {
...
- * Try to fallback to use UEFI removable path if the shim path doesn't exist.
- */
-#define FU_UEFI_DEVICE_FLAG_FALLBACK_TO_REMOVABLE_PATH (1 << 4)
 /**
  * FU_UEFI_DEVICE_FLAG_USE_SHIM_FOR_SB:
  *
  * Use shim to load fwupdx64.efi when SecureBoot is turned on.
  */
-#define FU_UEFI_DEVICE_FLAG_USE_SHIM_FOR_SB (1 << 5)
+#define FU_UEFI_DEVICE_FLAG_USE_SHIM_FOR_SB (1 << 4)
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-uefi-device.h:74]: I think these can stay - a "hole" is fine

File: plugins/uefi-capsule/fu-uefi-common.c [hunk 2]
```diff
@@ -86,7 +86,7 @@ fu_uefi_get_esp_app_path(FuDevice *device, const gchar *esp_path, const gchar *c
 }
 
 gchar *
-fu_uefi_get_built_app_path(GError **error)
+fu_uefi_get_built_app_path(const gchar *binary, GError **error)
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-uefi-common.c:89]: can you add a comment to this function please -- for our future selves. Otherwise LGTM.

**Review by hughsie — APPROVED** [review]

**Comment by superm1** [comment]
@valentindavid I had a try with the artifacts produced from CI.  It seems that the apparmor rules would need to be changed for this if this is the direction we go.

This is what I get:
```
Failed to copy /snap/fwupd/x1/libexec/fwupd/efi/shimx64.efi.signed to /boot/efi/EFI/fwupd/shimx64.efi: Error opening file ?/boot/efi/EFI/fwupd/shimx64.efi?: Permission denied
```

**Comment by valentindavid** [comment]
@superm1 Can you look at https://github.com/snapcore/snapd/pull/12557 and tell me if this is what you want? I have tested it, and it fixed it.

Also I have noticed that fwupdmgr tries to reboot, but it does not have the permission. So I have opened https://github.com/fwupd/fwupd/pull/5513

**Comment by superm1** [comment]
> @superm1 Can you look at https://github.com/snapcore/snapd/pull/12557 and tell me if this is what you want? I have tested it, and it fixed it.
> 
> Also I have noticed that fwupdmgr tries to reboot, but it does not have the permission. So I have opened https://github.com/fwupd/fwupd/pull/5513

Yup, thanks!

And the os-release thing we did for previous solution is still useful, please don't back it out. We'll have more accurate reports when the snap is used and a user opts into reporting.

**Comment by xnox** [comment]
For Ubuntu, we have multiple shim signatures. To work on both Core and Classic Ubuntu it is best to use the signature that is signed by the Core interim key and chained to the Classic CA as it is trusted by all Ubuntu installations. I will propose a change to the snapcraft packaging.

**Comment by superm1** [comment]
> For Ubuntu, we have multiple shim signatures. To work on both Core and Classic Ubuntu it is best to use the signature that is signed by the Core interim key and chained to the Classic CA as it is trusted by all Ubuntu installations. I will propose a change to the snapcraft packaging.

I wasn't aware there were different signed binaries.

Could you do it "soon"? We're targeting the next release on Feb 23.

**Comment by valentindavid** [comment]
@superm1 I have noticed this PR can brick. We are installing a shim that is potentially newer than in the distro. That may revoke old bootloaders (grub) and brick the current OS installation. The only way to get to boot it is to clear shim UEFI variables or manually update the bootloader offline.

**Comment by valentindavid** [comment]
One way to work around this is to not take the latest shim from `shim-signed`. Maybe we can provide an older shim from a PPA. And  we update this shim only when all major distros have updated the latest shim.

**Comment by valentindavid** [comment]
Rejecting the update would be fine.

We could query UEFI variable `SbatLevelRT` to check what has been already revoked. And if the revocations do not match what we expect, this is when we should bock.

**Comment by superm1** [comment]
How do you know what you expect though?

**Comment by valentindavid** [comment]
We can read section `.sbatlevel` of the shim we want to install. https://github.com/rhboot/shim/blob/main/sbat_var.S
Then we should have at least `SBAT_VAR_PREVIOUS`, I think.

**Pull request merged**